include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,465 @@
+     */
+    std::shared_ptr<const BatchLinOpFactory> preconditioner{nullptr};
+
+    /**
+     * Already generated preconditioner. If one is provided, the factory
+     * `preconditioner` will be ignored.
+     */
+    std::shared_ptr<const BatchLinOp> generated_preconditioner{nullptr};
+};
+
+
+template <typename Parameters, typename Factory>
+struct enable_preconditioned_iterative_solver_factory_parameters
+    : enable_parameters_type<Parameters, Factory>,
+      preconditioned_iterative_solver_factory_parameters {
+    /**
+     * Default maximum number iterations allowed.
+     *
+     * Generated solvers are initialized with this value for their maximum
+     * iterations.
+     */
+    int GKO_FACTORY_PARAMETER_SCALAR(max_iterations, 100);
+
+    /**
+     * Default residual tolerance.
+     *
+     * Generated solvers are initialized with this value for their residual
+     * tolerance.
+     */
+    double GKO_FACTORY_PARAMETER_SCALAR(tolerance, 1e-11);
+
+    /**
+     * To specify which type of tolerance check is to be considered, absolute or
+     * relative (to the rhs l2 norm)
+     */
+    ::gko::batch::stop::tolerance_type GKO_FACTORY_PARAMETER_SCALAR(
+        tolerance_type, ::gko::batch::stop::tolerance_type::absolute);
+
+    /**
+     * Provides a preconditioner factory to be used by the iterative solver in a
+     * fluent interface.
+     * @see preconditioned_iterative_solver_factory_parameters::preconditioner
+     */
+    Parameters& with_preconditioner(
+        deferred_factory_parameter<BatchLinOpFactory> preconditioner)
+    {
+        this->preconditioner_generator = std::move(preconditioner);
+        this->deferred_factories["preconditioner"] = [](const auto& exec,
+                                                        auto& params) {
+            if (!params.preconditioner_generator.is_empty()) {
+                params.preconditioner =
+                    params.preconditioner_generator.on(exec);
+            }
+        };
+        return *self();
+    }
+
+    /**
+     * Provides a concrete preconditioner to be used by the iterative solver in
+     * a fluent interface.
+     * @see preconditioned_iterative_solver_factory_parameters::preconditioner
+     */
+    Parameters& with_generated_preconditioner(
+        std::shared_ptr<const BatchLinOp> generated_preconditioner)
+    {
+        this->generated_preconditioner = std::move(generated_preconditioner);
+        return *self();
+    }
+
+private:
+    GKO_ENABLE_SELF(Parameters);
+
+    deferred_factory_parameter<BatchLinOpFactory> preconditioner_generator;
+};
+
+
+/**
+ * This mixin provides apply and common iterative solver functionality to all
+ * the batched solvers.
+ *
+ * @tparam ConcreteSolver  The concrete solver class.
+ * @tparam ValueType  The value type of the multivectors.
+ * @tparam PolymorphicBase  The base class; must be a subclass of BatchLinOp.
+ */
+template <typename ConcreteSolver, typename ValueType,
```

Comment:
the issue disallowing `using ValueType = typename ConcreateSolver::...` is related to https://stackoverflow.com/questions/652155/invalid-use-of-incomplete-type
I do not think adding another class trick mentioned in link is better than another ValueType template.
For future, I guess we can make `ValueType = void` to adapt base class of Multivector or varadic type without breaking interface.